src/hotspot/share/cds/metaspaceShared.cpp

```diff
@@ -1039,6 +1048,132 @@ bool MetaspaceShared::write_static_archive(ArchiveBuilder* builder, FileMapInfo*
   return true;
 }
 
+static void print_java_launcher(outputStream* st) {
+  st->print("%s%sbin%sjava", Arguments::get_java_home(), os::file_separator(), os::file_separator());
+}
+
+static void append_args(GrowableArray<Handle>* args, const char* arg, TRAPS) {
+  Handle string = java_lang_String::create_from_str(arg, CHECK);
+  args->append(string);
+}
+
+// Pass all options in Arguments::jvm_args_array() to a child JVM process
+// using the JAVA_TOOL_OPTIONS environment variable.
+static int exec_jvm_with_java_tool_options(const char* java_launcher_path, TRAPS) {
+  ResourceMark rm(THREAD);
+  HandleMark hm(THREAD);
+  GrowableArray<Handle> args;
+
+  const char* cp = Arguments::get_appclasspath();
+  if (cp != nullptr && strlen(cp) > 0 && strcmp(cp, ".") != 0) {
+    // We cannot use "-cp", because "-cp" is only interpreted by the java launcher,
+    // and is not interpreter by arguments.cpp when it loads args from JAVA_TOOL_OPTIONS
+    stringStream ss;
+    ss.print("-Djava.class.path=");
+    ss.print_raw(cp);
+    append_args(&args, ss.freeze(), CHECK_0);
+    // CDS$ProcessLauncher::execWithJavaToolOptions() must unset CLASSPATH, which has
+    // a higher priority than -Djava.class.path=
+  }
+
+  // Pass all arguments. These include those from JAVA_TOOL_OPTIONS and _JAVA_OPTIONS.
+  for (int i = 0; i < Arguments::num_jvm_args(); i++) {
+    const char* arg = Arguments::jvm_args_array()[i];
+    if (strncmp("-XX:AOTCacheOutput=", arg, 19) == 0 ||
```

Comment:
I have seen this hard coding of string length in the `strncmp` call many times in the code base. I wonder why we don't use something like this:
```
strncmp("-XX:AOTCacheOutput=", arg, sizeof("-XX:AOTCacheOutput=")-1)
```
Counting chars is really error prone. sizeof("...")-1 gives the same result.

Reply:
I end up using `strstr()`, as I don't want to repeat the string literal twice (risking a typo) or use a macro.